# The Python Challenge 

## Level 13 - call him

### Url: http://huge:file@www.pythonchallenge.com/pc/return/disproportional.html

纸面提示： call him, phone that evil，图中间5号键能点击，点击会转入`phonebook.php`:
```php
<methodResponse>
    <fault>
        <value>
            <struct>
                <member>
                    <name>faultCode</name>
                    <value>
                        <int>105</int>
                    </value>
                </member>
                <member>
                    <name>faultString</name>
                    <value>
                    <string>XML error 5: empty document</string>
                    </value>
                </member>
            </struct>
        </value>
    </fault>
</methodResponse>
```
并有附注：
> This XML file does not appear to have any style information associated with it. The document tree is shown below.
> 
> 这个XML文件似乎没有任何与之相关的样式信息。文件树显示如下。*（如上xd）*

[官方论坛](https://groups.google.com/g/python-challenge/search?q=L13)中有作者天语：

1. The root node has a very special name, google it!
2. Q: I am getting "He is not the evil", not sure who is he ?
   
   A: Look around in level 12

根据提示，使用 `methodResponse python` 关键词搜索，发现了[python的xmlrpc.client库](https://docs.python.org/zh-cn/3/library/xmlrpc.client.html)。

其中 Fault 对象的文档长这个样子：

> class xmlrpc.client.Fault
> 
>   Fault 对象封装了 XML-RPC fault 标签的内容。 Fault 对象具有下列属性:
> 
>   faultCode
> 
>       一个指明 fault 类型的整数。
> 
>   faultString
> 
>       一个包含与 fault 相关联的诊断消息的字符串。

长得很像上面php文件的样子，怀疑此模块即为突破口。

In [1]:
# 文档里还提到了ServerProxy.system.listMethods()这个方法，可以显示该服务器支持的方法，所以：

import xmlrpc.client
with xmlrpc.client.ServerProxy("http://www.pythonchallenge.com/pc/phonebook.php") as proxy:
    print(proxy.system.listMethods())


['phone', 'system.listMethods', 'system.methodHelp', 'system.methodSignature', 'system.multicall', 'system.getCapabilities']


出现了！这个叫 phone 的申必方法！接下来就要考虑如何使用这个方法了。

In [2]:
# 文档里还还提到了ServerProxy.system.methodHelp(name)这个方法，返回描述相应方法用法的文档字符串，所以：

import xmlrpc.client
with xmlrpc.client.ServerProxy("http://www.pythonchallenge.com/pc/phonebook.php") as proxy:
    for method in proxy.system.listMethods():
        print(method, ":", proxy.system.methodHelp(method))


phone : Returns the phone of a person
system.listMethods : This method lists all the methods that the XML-RPC server knows how to dispatch
system.methodHelp : Returns help text if defined for the method passed, otherwise returns an empty string
system.methodSignature : Returns an array of known signatures (an array of arrays) for the method name passed. If no signatures are known, returns a none-array (test for type != array to detect missing signature)
system.multicall : Boxcar multiple RPC calls in one request. See http://www.xmlrpc.com/discuss/msgReader$1208 for details
system.getCapabilities : This method lists all the capabilities that the XML-RPC server has: the (more or less standard) extensions to the xmlrpc spec that it adheres to


那么phone方法里要填谁呢？

第一反应先试试evil。

In [3]:
import xmlrpc.client
with xmlrpc.client.ServerProxy("http://www.pythonchallenge.com/pc/phonebook.php") as proxy:
    print(proxy.phone('evil'))

He is not the evil


？？？

答案是：在上一关我们查到evil4后就没查了，但恰恰是evil4这里藏有玄机。http://huge:file@www.pythonchallenge.com/pc/return/evil4.jpg 打开这个页面，F12看审查元素，并且切到Sources一栏而非Elements！

你就可以看到 evil4.jpg 这个文件里写着：

> Bert is evil! go back!

In [4]:
import xmlrpc.client
with xmlrpc.client.ServerProxy("http://www.pythonchallenge.com/pc/phonebook.php") as proxy:
    print(proxy.phone('Bert'))


555-ITALY


更搞的来了，尝试 http://www.pythonchallenge.com/pc/return/555-ITALY.html 这个页面居然是404。

不过这个不是问题，这是外国电话的老传统了，即先选定一个代表本单位的单词，然后在9键键盘上找到对应的数字键。

比如这里 555-ITALY 就对应着 555-48259 。

但是问题叒来了，不管是555-ITALY还是555-48259还是55548259都不对。

最后想到url好歹得是英文，所以试了一把ITALY。

回显说 SMALL letters.

## Next Level:

### Url: http://www.huge:file@pythonchallenge.com/pc/return/italy.html